staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -979,6 +994,34 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	return allErrs
 }
 
+func getCRDCost(baseCost uint64, schemaNode *schemaTree) uint64 {
+	totalCost := baseCost
+	parentNode := schemaNode.Parent
+	for parentNode != nil {
+		parentSchema := parentNode.Schema
+		if parentSchema.Type != "array" {
+			parentNode = parentNode.Parent
+			continue
+		} else if parentNode.MaxLength == 0 {
+			// we haven't calculated a max length for this entry in the map yet
+			if parentSchema.MaxLength != nil {
+				parentNode.MaxLength = int64(*parentSchema.MaxLength)
+			} else {
+				structural, err := structuralschema.NewStructural(parentSchema)
+				if err != nil {
+					// TODO(DangerOnTheRanger): how to add better error handling?
+					return totalCost
+				}
+				celSchema := celschema.SchemaDeclType(structural, false)
```

Comment:
`structuralschema.NewStructural` and `celschema.SchemaDeclType` can be quite expensive, particularly for schema nodes near the root.